### PR TITLE
Initial Multiaddr Implementation

### PR DESCRIPTION
**Pull request by aria42**
There are more protocols left to implement from [the partial spec](https://multiformats.io/multiaddr/), but thought this would be a good point to check in and get feedback, since the core pieces are here. 

I have the `Address` object support a collection view on the pairs of `[protocol-key value]`, which seems like the most natural way to use this data (e.g, configuring a network client). 

## Reading Guide

*  Start with `address.clj` to understand `Address` object and choice to have it support basic collection view of `[protocol-key value]` pairs. 
* `address/codec.clj` to understand `Codec`; the IP address implementation in particular was annoying since built-in platform libraries, specially on JS, almost have what you want, but require a decent amount of futzing. 

## Questions

* Some of the generic byte array manipulation seems generic and useful, should they be in `alphabase`?
* The `Codec` abstraction I'm using I think is needed since there's a few distinct formats for network protocols (e.g, DNS, IP, Port numbers, Onion, etc.). Should the interface be lower-level to use `offset` and `length` to allow direct manipulation of a buffer and avoid some byte array copies?

## TODO 
- [] Add UTF8 address codec for DNS protocol
- [] Add custom onion protocol 
- [] More tests, particularly on validation.

### CODE REVIEW

File: src/multiformats/address.cljc
Comment (greglook) [inline src/multiformats/address.cljc]: Indent is a bit off here, but I'll probably be running `cljfmt` over this anyway.

File: src/multiformats/address.cljc
Comment (greglook) [inline src/multiformats/address.cljc]: Destructuring doesn't seem necessary here:
```suggestion
               codec (:codec (ensure-protocol-attrs protocol-key))
```

File: src/multiformats/address.cljc
Comment (greglook) [inline src/multiformats/address.cljc:130]: Feels like this could be more efficient by expanding the conditional tree, instead of allocating zero-length byte arrays. On the other hand, it's also not likely to be on any hot paths anytime soon.

File: src/multiformats/address.cljc
Comment (greglook) [inline src/multiformats/address.cljc]: I believe that standard behavior is to preserve metadata on emptied collections:
```clojure
=> (-> {} (with-meta {:foo 123}) empty meta)
{:foo 123}
```

File: src/multiformats/address.cljc
Comment (greglook) [inline src/multiformats/address.cljc:151]: TIL Clojure can also use the hyphen-prefixed field access!

File: src/multiformats/address.cljc
Comment (greglook) [inline src/multiformats/address.cljc]: You could collapse the `let` above and build the parts list in this loop binding.

File: src/multiformats/address.cljc
Comment (greglook) [inline src/multiformats/address.cljc]: When is `(second parts)` not a string? I suppose this doesn't preclude an incomplete input like `/ip4`.

File: src/multiformats/address.cljc
Comment (greglook) [inline src/multiformats/address.cljc]: Could use `(nnext parts)` for concision and eagerness.

File: src/multiformats/address.cljc
Comment (greglook) [inline src/multiformats/address.cljc]: A more efficient approach here would be to directly encode the entries into a byte array before constructing the Address type. Without special reduction protocols, it seems like this is going to allocate O(n^2) memory.

File: src/multiformats/address/codec.cljc
Comment (greglook) [inline src/multiformats/address/codec.cljc:150]: Ah, some of the zero-byte array stuff above makes more sense with these checks.

File: test/multiformats/address/codec_test.cljc
Comment (greglook) [inline test/multiformats/address/codec_test.cljc]: 🏌️ `b/init-bytes`.

**Review by greglook — COMMENTED** [review]
Nice, thanks for tackling this! The spec is a little under-developed compared to the other multiformats, but it looks like you've got most of the useful pieces here. I left a number of comments and a couple of questions.

**Comment by greglook** [comment]
> Some of the generic byte array manipulation seems generic and useful, should they be in alphabase?

Yeah, the slice-copy seems useful, as does concat-arrays. At this point alphabase is more useful as a byte-manipulation library than as a base-encoder. 😅

> The Codec abstraction I'm using I think is needed since there's a few distinct formats for network protocols (e.g, DNS, IP, Port numbers, Onion, etc.). Should the interface be lower-level to use offset and length to allow direct manipulation of a buffer and avoid some byte array copies?

A few of my comments were along these lines - since the codecs are an implementation detail, I think it'd be okay to ship this as an initial version and optimize more later. If you feel like tackling this though, I'm all for it.

File: src/multiformats/address.cljc
Comment (aria42) [reply src/multiformats/address.cljc]: This might be unnecessary, but wanted to allow for integers (since some of these arguments are for ports and things that are naturally ints) and anything else that could be coerced to string this way, but happy to remove.

But I did update this to throw if we don't get a value for a protocol expecting one since I hadn't covered that.

**Comment by aria42** [comment]
Ok, I addressed comments and changed just about everything you suggested, I'll make a separate PR in `alphabase` with versions of these functions, but leave these here for now and remove once that's merged.

File: src/multiformats/address/codec.cljc
Comment (aria42) [reply src/multiformats/address/codec.cljc:150]: Ok, I actually left this in place defensively, but changed the call-site code to not bother creating 0 byte arrays

File: src/multiformats/address.cljc
Comment (aria42) [reply src/multiformats/address.cljc]: Done, now this just calls `(map entry-bytes)` and then concatenates results so memory is now O(n)

File: src/multiformats/address.cljc
Comment (aria42) [reply src/multiformats/address.cljc]: Ran `lein cljfmt fix` as its own commit.

File: src/multiformats/address.cljc
Comment (greglook) [inline src/multiformats/address.cljc]: I think I'll update alphabase so that this function ignores nils, which would make this form simpler:
```clojure
(b/concat code-bytes val-len-bytes val-bytes)
```

File: src/multiformats/address.cljc
Comment (greglook) [inline src/multiformats/address.cljc]: Ah, I'm guessing this is the default cljfmt? This should only be indented one level.

File: src/multiformats/base.cljc
Comment (greglook) [inline src/multiformats/base.cljc]: The other namespaces are already styled correctly, so I'd prefer you didn't reformat them.

**Review by greglook — APPROVED** [review]
Offset changes look good! I'll merge once you revert the changes to the other namespaces.

File: src/multiformats/address.cljc
Comment (aria42) [reply src/multiformats/address.cljc]: Oh weird, sorry out of the box it treats `deftype` as another function instead of a `def`-style form.

**Comment by aria42** [comment]
@greglook   I'm not sure why the CircleCI test is failing for cljs; I've got the latest `rhino` and `phantomjs` which appears to be what the test is using for dependencies. Locally, I'm running without error, but don't have a way to repro without getting access to CircleCI environment. 
screen shot 2018-12-25 at 9 21 27 am

**Comment by greglook** [comment]
Appears to be 5 instances of this error, three in the uint test and two in the utf8 test:
```
ERROR in (uint-codec-test) (at target/cljs/tests.js:NaN:4582)
round trip on numbers
expected: (= "80" (roundtrip codec/utf8-codec "80"))
  actual: #object[TypeError TypeError: second argument to Function.prototype.apply must be an array]
```

**Comment by aria42** [comment]
Oh I see the error in CircleCI, I just can't reproduce since tests passed on macOS with a clean build and the latest rhino; I'll add circleci to my fork so I can ssh into the build container and figure out what's going on. Alternatively, if you grant me admin on CircleCI I can do the same on your instance.

**Comment by aria42** [comment]
Ok so the issue is the version of `rhino` that CircleCI pulls in from Linux doesn't seem to have Typed Arrays and the underlying failure is that my UTF8 codec needs `Uint8Array` which doesn't exist on Rhino without some configuration that's non-trivial to pass in. I'm surprised this only broke on the recent push (was Rhino updated?) since all of my builds should've triggered this.
screen shot 2018-12-25 at 1 33 33 pm

If I replace the test to use `phantomjs` instead of `rhino` on the CircleCI server there's no problem since phantom seems to default to the language version that has typed arrays. 

So I can resolve this by changing the cljs test to use phantom environment instead. I'll put up a commit to do that and let you figure out how you want to deal.

**Comment by greglook** [comment]
🎉

**Pull request merged**
